include/deal.II/lac/qr.h

```diff
@@ -0,0 +1,863 @@
+   */
+  void
+  apply_givens_rotation(const unsigned int i, const unsigned int k);
+
+  /**
+   * Temporary vector needed to do Givens rotation of Q.
+   */
+  VectorType tmp;
+};
+
+
+
+/**
+ * The class to obtain $R$ triangular matrix of $A=QR$ factorization
+ * together with the matrix $A$ itself. Orthonormal $Q$ is not stored
+ * explicitly, but can be obtained from ${R^T}^{-1}A^T = Q^T$, thus
+ * the name of the class.
+ *
+ * The class is designed to update a given (possibly empty) QR factorization
+ * due to the addition of a new column vector. This is equivalent to
+ * constructing an orthonormal basis by the Gram-Schmidt procedure.
+ * The class also provides update functionality when the column
+ * is removed.
+ *
+ * The VectorType can be both parallel and serial vector and only need
+ * to have basic operations such as additions, scalar product, etc.
+ * It also needs to have a copy-constructor.
+ */
+template <typename VectorType>
+class ImplicitQR : public BaseQR<VectorType>
+{
+public:
+  /**
+   * Number type for R matrix.
+   */
+  typedef typename VectorType::value_type Number;
+
+  /**
+   * Default constructor.
+   */
+  ImplicitQR();
+
+  /**
+   * Destructor.
+   */
+  virtual ~ImplicitQR() = default;
+
+  virtual bool
+  append_column(const VectorType &column);
+
+  /**
+   * Remove column and update QR factorization.
+   *
+   * Starting from the given QR decomposition
+   * $QR= A = [a_1\,\dots a_n], \quad a_i \in R^m$
+   * we aim at computing factorization of
+   * $\tilde Q \tilde R= \tilde A = [a_2\,\dots a_n], \quad a_i \in R^m$.
+   *
+   * Note that $\tilde R^T \tilde R = \tilde A^T \tilde A$,
+   * where the RHS is included in $A^T A = R^T R$. Therefore
+   * $\tilde R$ can be obtained by Cholesky decomposition.
+   */
+  virtual void
+  remove_column(const unsigned int k = 0);
+
+  virtual void
+  multiply_with_Q(VectorType &y, const Vector<Number> &x) const;
+
+  virtual void
+  multiply_with_QT(Vector<Number> &y, const VectorType &x) const;
+
+  virtual void
+  multiply_with_A(VectorType &y, const Vector<Number> &x) const;
+
+  virtual void
+  multiply_with_AT(Vector<Number> &y, const VectorType &x) const;
+
+  /**
+   * Connect a slot to implement a custom check of linear dependency
+   * during addition of a column.
+   */
+  boost::signals2::connection
+  connect_append_column_slot(
+    const std::function<bool(const Vector<Number> &u,
+                             const Number &        rho2,
+                             const Number &        col_norm_sqr)> &slot);
+
+private:
+  /**
+   * Apply givens rotation in plane @i @p k to zero out $R(k,k)$.
+   */
+  void
+  apply_givens_rotation(const unsigned int i, const unsigned int k);
+
+  /**
+   * Signal used to decide if the new column is linear dependent.
+   *
+   * Here, @p u is the last column of the to-be R matrix , @p rho
+   * is its diagonal and @p col_norm_sqr is the square of the $l2$ norm of the column.
```

Comment:
maybe copy this to the documentation of the function that allows attaching to the signal